pkg/controller/bundle_sync_task.go

```diff
@@ -42,7 +44,13 @@ type bundleSyncTask struct {
 // READY state might mean something different for each resource type. For a Custom Resource it may mean
 // that a field "State" in the Status of the resource is set to "Ready". It is customizable via
 // annotations with some defaults.
-func (st *bundleSyncTask) process() (retriableError bool, e error) {
+func (st *bundleSyncTask) processNormal() (retriableError bool, e error) {
+	// If the "deleteResources" finalizer is missing, add it and finish the processing iteration
+	if !hasDeleteResourcesFinalizer(st.bundle) {
+		st.newFinalizers = addDeleteResourcesFinalizer(st.bundle.GetFinalizers())
+		return false, nil
```

Comment:
Why do we need to early return here without checking the deleteTimeStamp exists?

Reply:
because we invoke this method only if `deletionTimestamp` is set, see controller_worker.go

Reply:
oops, sorry. `processNormal` invoked only if `deletionTimestamp` is **NOT** set. `processDeleted` is invoked otherwise.

Reply:
It should be safe to continue processing the bundle without early returning then no?

Reply:
@amckague no, it's not safe:
1. Smith reads a bundle. Adds a finalizer and starts creating resources immediately.
2. User deletes the bundle.
3. Smith finished creating the resources and tries to update the bundle (finalizers and status). It fails, because the version has changed at the step 2.
4. CRD apiserver sees the `deletionTimestamp` and no finalizers and proceeds the deletion.
5. Bundle is gone, but all resources are still there :(

Reply:
In other words, we need to store finalizer in etcd before we do any actions with resources.

Reply:
Interesting. I read this as trying hard to prevent users from removing the finalizer rather than just making sure we initialize properly.
Perhaps https://kubernetes.io/docs/admin/extensible-admission-controllers/#what-are-initializers would be closer to this idea.

Reply:
@amckague agreed that it's possible to:
- write a dynamic admission controller which adds an initializer
- in the initializer processing add a finalizer and remove the initializer
Not sure if it's worth it, but could be an improvement in the follow-up PR.